packages/evm/src/eof/container.ts

```diff
@@ -0,0 +1,226 @@
+      validationError(EOFError.VerifyBytes, this.ptr - expect.length, errorStr)
+    }
+  }
+
+  readUint(errorStr?: string) {
+    if (this.ptr >= this.data.length) {
+      validationError(EOFError.OutOfBounds, this.ptr, errorStr)
+    }
+    return this.data.readUint8(this.ptr++)
+  }
+
+  verifyUint(expect: number, errorStr?: string) {
+    if (this.readUint() !== expect) {
+      validationError(EOFError.VerifyUint, this.ptr - 1, errorStr)
+    }
+  }
+
+  readUint16(errorStr?: string) {
+    const end = this.ptr + 2
+    if (end > this.data.length) {
+      validationError(EOFError.OutOfBounds, this.ptr, errorStr)
+    }
+    const ptr = this.ptr
+    this.ptr += 2
+    return this.data.readUint16BE(ptr)
+  }
+
+  getPtr() {
+    return this.ptr
+  }
+
+  isAtEnd() {
+    return this.ptr === this.data.length
+  }
+}
+
+class EOFHeader {
+  typeSize: number
+  codeSizes: number[]
+  dataSize: number
+  buffer: Buffer
+
+  private codeStartPos: number[]
+
+  constructor(buf: Buffer) {
+    const stream = new StreamReader(buf)
+    stream.verifyBytes(MAGIC, EOFError.MAGIC)
+    stream.verifyUint(VERSION, EOFError.VERSION)
+    stream.verifyUint(KIND_TYPE, EOFError.KIND_TYPE)
+    const typeSize = stream.readUint16(EOFError.TypeSize)
+    if (typeSize < 4) {
+      validationError(EOFError.InvalidTypeSize, typeSize)
+    }
+    if (typeSize % 4 !== 0) {
+      validationError(EOFError.InvalidTypeSize, typeSize)
+    }
+    stream.verifyUint(KIND_CODE, EOFError.KIND_CODE)
+    const codeSize = stream.readUint16(EOFError.CodeSize)
+    if (codeSize < 1) {
+      validationError(EOFError.MinCodeSections)
+    }
+    if (codeSize > 1024) {
+      validationError(EOFError.MaxCodeSections)
+    }
+    const codeSizes = []
+    for (let i = 0; i < codeSize; i++) {
+      const codeSectionSize = stream.readUint16(EOFError.CodeSection)
+      if (codeSectionSize === 0) {
+        validationError(EOFError.CodeSectionSize)
+      }
+      codeSizes.push(codeSectionSize)
+    }
+    stream.verifyUint(KIND_DATA, EOFError.KIND_DATA)
+    const dataSize = stream.readUint16(EOFError.DataSize)
+    stream.verifyUint(TERMINATOR, EOFError.TERMINATOR)
+
+    this.typeSize = typeSize
+    this.codeSizes = codeSizes
+    this.dataSize = dataSize
+    this.buffer = buf.slice(0, stream.getPtr())
+    this.codeStartPos = [0]
+  }
+
+  sections() {
+    return [this.typeSize, this.codeSizes, this.dataSize]
+  }
+  sectionSizes() {
+    return [1, this.codeSizes.length, 1]
+  }
+
+  // Returns the code position in the code section
+  // This is thus not relative to the entire container, but works on the runtime code itself
+  getCodePosition(section: number) {
+    if (this.codeStartPos[section]) {
+      return this.codeStartPos[section]
+    }
+    const start = this.codeStartPos.length
+    let offset = this.codeStartPos[start - 1]
+    for (let i = start; i <= section; i++) {
+      offset += this.codeSizes[i - 1]
+      this.codeStartPos[i] = offset
```

Comment:
I think I must be missing the point of this function but why are you updating the `codeStartPosition` in this method?  As best I can tell, this method just returns the position in the overall code container of where a code section starts.  Is that correct?

Reply:
If you load the header, then it is unknown at what offset the code sections start. If you try to read section 2, then it will first see if we have it cached. Otherwise, it will start at the latest code offset which we know, and then iterate over the remaining code section sizes to figure out where the offset of the code section which we want is. And yes, it is indeed relative to the entire container (I should document this method)